rfcs/20240919-dnncompat/README.md

```diff
@@ -0,0 +1,936 @@
+their include path and adding `#include <dnncompat/dnncompat.hpp>` to their
+project and linking with the oneDNN library.
+
+## SYCLomatic Usage
+
+At present, SYCLomatic provides its own header-only wrapper for oneDNN and
+oneMKL which is used when converting applications. While this approach is
+sufficient, the code produced by SYCLomatic can sometimes be challenging to
+debug and understand for new oneDNN users as it relies on a mixture of custom
+wrappers around oneDNN types along with direct oneDNN API calls. Additionally,
+there are several differences between the supported primitives & their supported
+features in oneDNN and vendor-specific libraries, making debugging incorrect
+output more difficult.
+
+To address these issues, the SYCLomatic headers can be substituted with the
+`dnncompat` wrapper. By using `dnncompat`, the conversion process will be
+simplified, reducing the complexity of the generated code and making it more
+straightforward to understand and debug. This will ensure that developers can
+convert their applications to `dnncompat` & oneDNN more easily, while also
+reducing potential integration issues.
+
+## Scope
+
+Initially the focus will be on a specific set of operators commonly used in the
+VGG-16 model. The following operators will be prioritised: Convolution, BiasAdd,
+Pooling, Activation, Softmax, Local Response Normalization (LRN), and
+TransformTensor (transpose). This will allow for creating a relatively small
+prototype, while showing that `dnncompat` will be useful.
+
+In this initial implementation, these operators will be supported in the forward
+direction. The supported data types initially will be `f32` and `f16`. In the
+future, the support for operations, data types & directions can be expanded
+based on the need of SYCLomatic and community.
+
+## Implementation Details
+
+`dnncompat` will adopt an API design similar to vendor-specific libraries, which
+are often C-based. Tensors and operations will be described using descriptors
+(objects that encapsulate the necessary parameters for a given tensor or
+operation). These descriptors will then be passed, together with pointers to GPU
+ memory containing the tensor data, to the relevant API functions which will
+execute the operations based on the provided parameters.
+
+While `dnncompat` will be implemented with object-oriented programming (OOP)
+functionality as it is a C++ library, the user-facing API will primarily
+interact with these descriptors in a C-style API rather than through member
+functions of classes. Since `dnncompat` is a C++ library, OOP features can be
+used, but the C-style APIs will be the main user facing API to align with the
+design of vendor-specific libraries, such as `cuDNN`.
+
+The classes in `dnncompat` will be presented to users as opaque pointers, hiding
+implementation details and providing API as close as possible to vendor-specific
+libraries.
+
```

Comment:
(random spot) A few comments here:
- what is the plan for features that are not supported by oneDNN? Good examples are Nanpropagation flag or convolution algorithms. You mention below we should print warnings, but I failed to find if we return a non successful status, or if we move forward with execution? And if we move forward with execution, what should be the values used?
- what about scratchpad management? IIUC, temporary buffer management is exposed in cuDNN and might not map to how it works for oneDNN. e.g. what about when oneDNN underlying primitive require scratchpad, but cuDNN symbol does not? Shall we transparently allocate that scratchpad? And if so, shall we notify the user?

Reply:
- For cases where this will affect the accuracy of the output we could return non-success status. For cases where it does not affect the accuracy we could print the warning and continue execution. It could be decide on a case-by-case basis.
- That's a good question. There are 2 possible options, let me think about it and we can discuss which approach is better.